File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack();

        for (char c : s.toCharArray()) {
            if (c == '[' || c == '{' || c == '(') {
                stack.push(c);
            } else {
                if(stack.isEmpty()) {
                    return false;
                }
                char removed = stack.pop();
                if ((c == '}' && removed != '{') || (c == ']' && removed != '[') || (c == ')' && removed != '(')) {
                    return false;
                }
            }

        }
        return stack.isEmpty();
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.cs
public class Solution {
    private Dictionary<char, char> mapping;
    public Solution() {
        mapping = new Dictionary<char, char> {
            {'{', '}'}, {'[', ']'}, {'(', ')'};
        };
    }

    public bool IsValid(string s) {
        var stack = new Stack<char>();
        foreach(var c in s) {
            if(mappings.ContainsKey(c)) {
                char topElement = stack.Count == 0 ? '#' : stack.Pop();
                if(topElement != mappings[c]) {
                    return false;
                }
            }
            else {
                stack.Push(c);
            }
        }
        return stack.Count == 0;
    }
}